### ISSUE TITLE
How much iterations i need?

### ISSUE THREAD

**Issue by hadaev8** [NVIDIA/tacotron2#111]
Training on Russian dataset, output says I have less than 0.2 loss, the default is 500 epoch.
Now I'm on 1333 epoch and still get _Warning! Reached max decoder steps_
Should I counting or it screwed up?
http://puu.sh/CgXpt/41886048cd.jpg

**Comment by rafaelvalle** [comment]
Make sure the model has learned attention during training.
We have found that it is easier for the model to learn attention if silence is trimmed from the beginning and end of audio files.

**Comment by hadaev8** [comment]
I listened to the files, there is no (or too little) silence at the beginning or end in dataset.
I will keep in mind that it is better to remove silence, but probably not the best idea to change dataset during training (or its ok? bit scary to ruin everything couz im using tesla k80 and its all need a lot of time).
So how long does it take LJ Speech?
I have 0.2 val loss decrease by 1k steps, so it trains, but slower then I expected by 500 iterations in defult hyper params file.

Also do I need to train tacotron before waveglow? It seems strange to me that I can put waveglow training only with audio files without spectrograms from tacotron.

**Comment by rafaelvalle** [comment]
How big is the dataset?

**Comment by hadaev8** [comment]
Im using 1st male russian speaker from here
http://www.m-ailabs.bayern/en/the-mailabs-speech-dataset/
20 hours total

**Comment by rafaelvalle** [comment]
It is ok to train Tacotron before glow.
Can you share your current tensorboard with alignments?
If the model doesn't learn the alignments you can't perform inference.

**Comment by hadaev8** [comment]
I mean can I train tacotron and glow at the same time? By readme, it seems glow doesn't need tacotron model for train. It seems strange a bit.

My logs lost by crush of google colab vm, but I run tacotron from the checpoint against test set, seems hopeless.
https://colab.research.google.com/drive/1eGEXdrgioRs-7R0WOEx2_fHmDC-zDmzk#scrollTo=5nJcq9nHijEH&line=7&uniqifier=1

**Comment by hadaev8** [comment]
I can confirm that silence removal helps a lot to reduce loss.

**Comment by hadaev8** [comment]
Also here is logs
https://drive.google.com/open?id=1ugef68POLQXEO1ERSfW6qL_T0YEVbOIE
not sure how to interpret distributions and histograms.

**Comment by hadaev8** [comment]
Val loss
http://puu.sh/CiAio/87de1bd8c1.png
Alignment
http://puu.sh/CiAkW/eb0d821d1e.png
Mels
http://puu.sh/CiAlV/ad0f4f39af.png

**Comment by hadaev8** [comment]
I probably should think about it before, but need I put Russian symbols to text/symbols.py ?

**Comment by rafaelvalle** [comment]
Yes! Get the set of characters in the alphabet and modify symbols.py in the text/ folder.

**Comment by hadaev8** [comment]
As I understand it, there is not much difference, because one of the functions transfers characters to Latin.
Now I try Cyrillic characters in the dictionary and parameter base cleaners.

I still don’t understand how many iterations have to go before I can be sure that this doesn’t work.
When training for LJSpeech dataset, you reduced learning rate while training?

**Comment by rafaelvalle** [comment]
Train until you see a yellow-ish diagonal line on the attention plots.

**Comment by hadaev8** [comment]
But I am tormented by doubts that my dataset needs other hyper parameters.
How long does it take to be sure that the algorithm cannot converge?
If val loss started to increase, does that mean everything broke down?

**Comment by hadaev8** [comment]
Its again me
Step 4700
Seems it overfit
http://puu.sh/Cl8ls/e5bb692522.png
Attention doesnt work
http://puu.sh/Cl8m4/101a086539.jpg
Mels is bad too
http://puu.sh/Cl8n1/07e406388b.jpg

Any tips?

**Comment by hadaev8** [comment]
This is my default dataset https://drive.google.com/open?id=1FCwMelicl_dwE8Y5MwuGw7Pd4zDY-OVp
This is dataset with trimmed silence https://drive.google.com/open?id=13Mkp_6Sm_oj8jQoUvnYpxxZw6yXvcJpk

The code I'm used to preprocessing https://pastebin.com/4MzD1x6C
Spectrogramm, wavs predicted and from dataset https://drive.google.com/open?id=1e7Plu2x2WS8_k2yWa_ZJ_jWXGUU_oz01

I see predicted wav have 2 extra seconds, does it means I get this problem?

**Comment by Yeongtae** [comment]
Maybe yes, but it looks like that your model fail to learn attention values.

**Comment by hadaev8** [comment]
@Yeongtae 
On what iteration its attention starts looking ok?
What loss you have at the end?

**Comment by Yeongtae** [comment]
Maybe around 30 epoch.

**Comment by Yeongtae** [comment]
@hadaev8 
You must construct char-embedding function for your language.

**Comment by hadaev8** [comment]
Russian very similar to latin languages.
So, I'm just using basic cleaners with new alphabet symbols.
Seems everything is ok.
http://puu.sh/CtisY/131115d283.png

Also my current attention for 400 epoch, I can count it doesn't work, right?
http://puu.sh/CtiIa/59a17c577c.jpg

Also, I find CMUDict symbols greatly increase the number of characters in hyper param file.
Not sure if this parameter affects learning and what is CMUDict, but I will try to start training without it now.

**Comment by pravn** [comment]
What's the size of your dataset? Alignment won't be learnt unless the
corpus is large (~LJSpeech).

pravn.wordpress.com

**Comment by hadaev8** [comment]
20 hours, should be ok i think.

**Comment by Yeongtae** [comment]
Colud you share your dimension of char-embedding.

**Comment by hadaev8** [comment]
This, right?
http://puu.sh/CtkEZ/867200fd25.png
Without cmudict it is 45
https://github.com/hadaev8/tacotron2/blob/master/text/symbols.py

**Comment by Yeongtae** [comment]
![image](https://user-images.githubusercontent.com/13597229/50814461-1c592000-135d-11e9-9337-1f29d6099b53.png)
It's maybe helpful to reduce these parameters. ex) half
Because your char embedding dimension is so small.
In my case, 80.

**Comment by Yeongtae** [comment]
In addition, why are you use (sr:16000?)

**Comment by hadaev8** [comment]
I will try to reduce it.
Yes, it is 16000, its sr of dataset, is it important? I can resample it ofc.

**Comment by Yeongtae** [comment]
Yes, I recommend to you that you set 128 to encod-embed-dim

**Comment by hadaev8** [comment]
Should i reduce decoder_rnn_dim=1024 ?

**Comment by Yeongtae** [comment]
Not yet

**Comment by hadaev8** [comment]
128 128 encoder decoder dim, 32 batch size, rest default.
Still seems broken
http://puu.sh/CtIm8/5bbe8d1c05.png

**Comment by hadaev8** [comment]
@Yeongtae should you post your tensorflow logs? I wonna look how it looks then things works

**Comment by hadaev8** [comment]
Mm, no.
Now I'm trying with default sample rate, but it doesn't seems to coverage.
At least at 200 epoch http://puu.sh/Cv4LF/18cc92c80b.png

**Comment by Yeongtae** [comment]
If you dont see diagonal attention around 30 epoch, stop training

**Comment by hadaev8** [comment]
Im probably retarded.
Only now after all these days i come up with idea of checking files lists and find I mixed up file names and 9k samples was in validation file instead of train.
It should explain everything.

**Comment by Yeongtae** [comment]
In the studies related tacotron, phoneme embedding make more easy to converge model.
I recommend it to you.
Using lexicon model, make your own phone embedding function.

**Comment by hadaev8** [comment]
I can report it doesnt work with 128 and default 512 embedding/encoder dim
It how it looks now at 30 epoch for 512
http://puu.sh/Cz6hJ/f6a4ef05f8.png

Do i need phonem data for lexicon model?

**Comment by dnnnew** [comment]
@hadaev8 Any success?

**Comment by hadaev8** [comment]
No
attention looks this after 50 interations
http://puu.sh/CExPx/83a39b7259.png

Maybe I need more, idk.

**Comment by Yeongtae** [comment]
@hadaev8 
Normalizing values of melspectrogram can help to train model.
Because original values are skewed in negative direction.

Here is my implementation for normalization.
https://github.com/Yeongtae/tacotron2/blob/prosody_encoder_test/layers.py
https://github.com/Yeongtae/tacotron2/blob/prosody_encoder_test/inference.py
I refer to code of https://github.com/Rayhane-mamah/Tacotron-2

**Comment by hadaev8** [comment]
@Yeongtae 
Have you tried to reproduce the results?
33 iteration with default dataset and code.
Attention still doesnt work

**Comment by hadaev8** [comment]
Seems it work after 45 epoch.

**Comment by dnnnew** [comment]
@hadaev8 without any changes?

**Comment by hadaev8** [comment]
@dnnnew 
I think my problem is that I didn’t wait for enough time (30 epochs), now I’ll start and try 60

**Comment by hadaev8** [comment]
I noticed one thing
http://puu.sh/CS7Oa/78085e229c.png
http://puu.sh/CS7ON/cd3768c8f2.png

So it turns out the network is paying attention to silence?

**Comment by prajwalkr** [comment]
> If you dont see diagonal attention around 30 epoch, stop training

Could you please share your loss graph(s)? Thank you.

**Comment by rafaelvalle** [comment]
We have updated the README and code to make it easier to train a new model starting from our pre-trained model.This should be the best option when training a new model.

**Comment by maxbaluev** [comment]
@rafaelvalle Even if i need to teach the Russian language? Is it possible to train a model according to Russian data over an English model?

**Comment by NoahDrisort** [comment]
> In the studies related tacotron, phoneme embedding make more easy to converge model.
> I recommend it to you.
> Using lexicon model, make your own phoneme embedding function.

@Yeongtae Can I train with phoneme in this repo, can you share some highlight what I should do when using phoneme.

**Comment by Yeongtae** [comment]
@NoahDrisort 
Check this work https://github.com/Kyubyong/g2p

**Comment by CookiePPP** [comment]
@Yeongtae 
How well does g2p perform compared to a Forced Aligner?

**Comment by Yeongtae** [comment]
Check table1 in his work https://arxiv.org/pdf/1904.04169.pdf, but this work is voice conversion.

It is hard explaining how well. 
I think that g2p is just standard for text.

**Comment by EuphoriaCelestial** [comment]
> ![image](https://user-images.githubusercontent.com/13597229/50814461-1c592000-135d-11e9-9337-1f29d6099b53.png)
> It's maybe helpful to reduce these parameters. ex) half
> Because your char embedding dimension is so small.
> In my case, 80.

how should I adjust those param? I only have 101 symbols in total

**Comment by ErfolgreichCharismatisch** [comment]
Can you help me with https://github.com/NVIDIA/tacotron2/issues/463?
